admindb/sqlite/invites.go

```diff
@@ -0,0 +1,206 @@
+
+		inserted := false
+	trying:
+		for tries := 100; tries > 0; tries-- {
+			// generate an invite code
+			rand.Read(tokenBytes)
+
+			// hash the binary of the token for storage
+			h := sha256.New()
+			h.Write(tokenBytes)
+			newInvite.Token = fmt.Sprintf("%x", h.Sum(nil))
+
+			// insert the new invite
+			err := newInvite.Insert(ctx, tx, boil.Infer())
+			if err != nil {
+				var sqlErr sqlite3.Error
+				if errors.As(err, &sqlErr) && sqlErr.ExtendedCode == sqlite3.ErrConstraintUnique {
+					// generated an existing token, retry
+					continue trying
+				}
+				return err
+			}
+			inserted = true
+			break // no error means it worked!
+		}
+
+		if !inserted {
+			return errors.New("admindb: failed to generate an invite token in a reasonable amount of time")
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return "", err
+	}
+
+	return base64.URLEncoding.EncodeToString(tokenBytes), nil
+}
+
+// Consume checks if the passed token is still valid. If it is it adds newMember to the members of the room and invalidates the token.
+// If the token isn't valid, it returns an error.
+// Tokens need to be base64 URL encoded and when decoded be of tokenLength.
+func (i Invites) Consume(ctx context.Context, token string, newMember refs.FeedRef) (admindb.Invite, error) {
+	var inv admindb.Invite
+
+	tokenBytes, err := base64.URLEncoding.DecodeString(token)
+	if err != nil {
+		return inv, err
+	}
+
+	if n := len(tokenBytes); n != tokenLength {
+ 		return inv, fmt.Errorf("admindb: invalid invite token length (only got %d bytes)", n)
+	}
+
+	// hash the binary of the passed token
+	h := sha256.New()
+	h.Write(tokenBytes)
+	hashedToken := fmt.Sprintf("%x", h.Sum(nil))
+
+	err = transact(i.db, func(tx *sql.Tx) error {
+		entry, err := models.Invites(
+			qm.Where("active = true AND token = ?", hashedToken),
+			qm.Load("CreatedByAuthFallback"),
+		).One(ctx, tx)
+		if err != nil {
+			return err
+		}
+
+		err = i.allowList.add(ctx, tx, newMember)
+		if err != nil {
+			return err
+		}
+
+		entry.Active = false
+		_, err = entry.Update(ctx, tx, boil.Whitelist("active"))
+		if err != nil {
+			return err
+		}
+
+		inv.ID = entry.ID
+
+		if entry.AliasSuggestion.Valid {
+			inv.AliasSuggestion = *entry.AliasSuggestion.Ptr()
+		}
+
+		inv.CreatedBy.ID = entry.R.CreatedByAuthFallback.ID
+		inv.CreatedBy.Name = entry.R.CreatedByAuthFallback.Name
+
+		return nil
+	})
+
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return inv, admindb.ErrNotFound
+		}
+		return inv, err
+	}
+
+	return inv, nil
+}
+
+// List returns a list of all the valid invites
+func (i Invites) List(ctx context.Context) ([]admindb.Invite, error) {
+	var invs []admindb.Invite
+
+	err := transact(i.db, func(tx *sql.Tx) error {
+		entries, err := models.Invites(
+			qm.Where("active = true"),
+			qm.Load("CreatedByAuthFallback"),
+		).All(ctx, tx)
+		if err != nil {
+			return err
+		}
+
+		invs = make([]admindb.Invite, len(entries))
+		for i, e := range entries {
```

Comment:
overloading `i Invites` with `i` range count

Reply:
yikes. yea, that's bad. 

I just checked and noticed the other sqlite sub-service also have inconsistent names.

How about having `s` for service as the last letter?

So
* `func (is Invites) ...`
* `func (als AllowList) ...`
* `func (afs AuthFallback) ..`
* `func (as AliasService) ...`
* `func (ns Notes) ...`


Similar to the package name change I'd prefer this as a follow up, though.

Reply:
oops didn't respond: i really like the idea of keeping things cohesive! i'm ambivalent on using `s`, i don't really know the code well enough yet to have a feeling for how it works. some weird semantic clashes with `is`, `as`, `ns` but :shrug: